15-greenhouse-monitor/firmware/greenhouse/greenhouse.ino

```diff
@@ -0,0 +1,628 @@
+    setTaskIntervalFromEnvironment(env, changed, errors, &readSensorsInterval, "monitor_secs", 1000, DEFAULT_POLL_SENSORS_INTERVAL);
+    setTaskIntervalFromEnvironment(env, changed, errors, &sendReportInterval, "report_mins", 1000*60, DEFAULT_REPORT_INTERVAL);
+
+    updateAlertThresholdsFromEnvironment(env, changed, errors, DATA_FIELD_SOIL_TEMP, &soil_temp_intervals);
+    updateAlertThresholdsFromEnvironment(env, changed, errors, DATA_FIELD_SOIL_MOISTURE, &soil_moisture_intervals);
+
+    updateAlertThresholdsFromEnvironment(env, changed, errors, DATA_FIELD_AIR_TEMP, &air_temp_intervals);
+    updateAlertThresholdsFromEnvironment(env, changed, errors, DATA_FIELD_AIR_HUMIDITY, &air_humidity_intervals);
+    updateAlertThresholdsFromEnvironment(env, changed, errors, DATA_FIELD_AIR_PRESSURE, &air_pressure_intervals);
+
+    updateAlertThresholdsFromEnvironment(env, changed, errors, DATA_FIELD_LIGHT_LEVEL, &light_level_intervals);
+
+    char buf[256];
+    debug.println("environment updates:");
+    if (changed->child) {
+        JPrintPreallocated(changed, buf, sizeof(buf), false);
+        debug.println(buf);
+    }
+    else {
+        debug.println("none");
+    }
+    debug.println("environment errors:");
+    if (errors->child) {
+        JPrintPreallocated(errors, buf, sizeof(buf), false);
+        debug.println(buf);
+    }
+    else {
+        debug.println("none");
+    }
+    debug.println();
+
+    if (changed->child || errors->child) {
+        J* req = notecard.newRequest("note.add");
+        JAddStringToObject(req, "file", DATA_FILE_NOTIFY);
+        J* body = NoteNewBody();
+        if (changed->child) {
+            JAddItemToObject(body, "updates", changed);
+            changed = nullptr;
+        }
+        if (errors->child) {
+            JAddItemToObject(body, "errors", errors);
+            errors = nullptr;
+        }
+        JAddStringToObject(body, DATA_FIELD_APP, APP_NAME);
+        NoteAddBodyToObject(req, body);
+        notecard.sendRequest(req);
+    }
+    JDelete(changed);
+    JDelete(errors);
+}
+
+bool readEnvironment() {
+    // Read all env vars from the notecard in one transaction
+    J *rsp = notecard.requestAndResponse(notecard.newRequest("env.get"));
+    if (rsp == NULL) {
+        return false;
+    }
+    if (notecard.responseError(rsp)) {
+        notecard.deleteResponse(rsp);
+        return false;
+    }
+
+    // Update the env modified time
+    environmentModifiedTime = JGetNumber(rsp, "time");
+
+    // Update the environment
+    J *body = JGetObject(rsp, "body");
+    if (body != NULL) {
+        environmentUpdated(body);
+    }
+
+    // Done
+    notecard.deleteResponse(rsp);
+    return true;
+}
+
+void setup()
+{
+    // Initialize debug IO
+    pinMode(LED_BUILTIN, OUTPUT);
+    // uncomment to have the app wait for a Serial connection
+    while (!debug) {};
+    debug.begin(115200);
+    debug.println("*** " __DATE__ " " __TIME__ " ***");
+
+    // Initialize I2C
+    Wire.begin();
```

Comment:
Learned from @zfields recently that [this isn't required](https://github.com/blues/notecard-env-var-manager/pull/5#discussion_r1174581714).